service/src/DCGMDevicePool.cpp

```diff
@@ -0,0 +1,186 @@
+        return instance;
+    }
+
+    DCGMDevicePoolImp::DCGMDevicePoolImp()
+        : m_update_freq(1000)    // 1 millisecond
+        , m_max_keep_age(1.0)    // 1 second
+        , m_max_keep_sample(100) // 100 samples
+    {
+        // Guarantee geopm_dcgm_field_ids_e order is used
+        m_dcgm_field_ids[M_SM_ACTIVE] = DCGM_FI_PROF_SM_ACTIVE;
+        m_dcgm_field_ids[M_SM_OCCUPANCY] = DCGM_FI_PROF_SM_OCCUPANCY;
+        m_dcgm_field_ids[M_DRAM_ACTIVE] = DCGM_FI_PROF_DRAM_ACTIVE;
+
+        dcgmReturn_t result;
+
+        //Initialize DCGM
+        result = dcgmInit();
+        check_dcgm_result(result, "Error Initializing DCGM.", __LINE__);
+
+        // We are assuming a local version of DCGM.  This could transition to a
+        // dcgmStartEmbedded at a later date.
+        char *host_ip = "127.0.0.1";
+        result = dcgmConnect(host_ip, &m_dcgm_handle);
+        check_dcgm_result(result, "Error connecting to standalone DCGM instance", __LINE__);
+
+        //Check all devices are DCGM enabled
+        unsigned int dcgm_dev_id_list[DCGM_MAX_NUM_DEVICES];
+        result = dcgmGetAllSupportedDevices(m_dcgm_handle, dcgm_dev_id_list, &m_dcgm_dev_count);
+        check_dcgm_result(result, "Error fetching DCGM supported devices.", __LINE__);
+
+        for (int i = 0; i < m_dcgm_dev_count; ++i) {
+            std::vector<dcgmFieldValue_v1> field_values(sizeof(m_dcgm_field_ids)/sizeof(m_dcgm_field_ids[0]));
+            m_dcgm_field_values.push_back(field_values);
+        }
+
+        //Setup Field Group
+
+        char geopm_group[] = "geopm_field_group";
+        result = dcgmFieldGroupCreate(m_dcgm_handle, sizeof(m_dcgm_field_ids)/sizeof(m_dcgm_field_ids[0]), &m_dcgm_field_ids[0],
+                                      geopm_group, &m_field_group_id);
+
+        //Retry case
+        if (result == DCGM_ST_DUPLICATE_KEY) {
+#ifdef GEOPM_DEBUG
+            std::cerr << "DCGMDevicePool::" << std::string(__func__)  <<
+                         ": Duplicate field group found. " << std::endl;
+#endif
+            result = dcgmFieldGroupDestroy(m_dcgm_handle, m_field_group_id);
+            check_dcgm_result(result, "Error destroying DCGM geopm_fields group.", __LINE__);
+            result = dcgmFieldGroupCreate(m_dcgm_handle, sizeof(m_dcgm_field_ids)/sizeof(m_dcgm_field_ids[0]), &m_dcgm_field_ids[0],
+                                          geopm_group, &m_field_group_id);
+            check_dcgm_result(result, "Error re-creating DCGM geopm_fields group.", __LINE__);
+        }
+        else {
+            check_dcgm_result(result, "Error creating DCGM geopm_fields group.", __LINE__);
+        }
+
+        // Note: Currently we are using dcgmWatchFields, but may transition to using
+        //       dcgmProfWatchFields at a later date
+        result = dcgmWatchFields(m_dcgm_handle, DCGM_GROUP_ALL_GPUS, m_field_group_id, m_update_freq,
+                                 m_max_keep_age, m_max_keep_sample);
```

Comment:
If I understand the docs, I think this is requesting that DCGM cache up to m_max_keep_sample (100) most-recent values, internally sampled every 1000 usec, keeping up to m_max_keep_age (1) seconds (i.e., up to 1000, were it not for the explicit sample limit).

Does this mean that each call to `dcgmGetLatestValuesForFields` is ignoring 99 samples from each call to DCGM? Could we alternatively get sub-geopm-sample-rate information by using `dcgmGetValuesSince_v2` that gets an array of dcgm samples that were observed between geopm requests? This could be useful for observing spikes in non-monotonically-counting signals, like the utilization signals. In that case, we could report the average or peak utilization between geopm samples instead of the point-in-time utilization when geopm gets its samples.

The docs aren't clear if that's the difference between those functions in their long-form descriptions. This is all just assumptions based on their names.

Reply:
In testing using dcgmGetValuesSince wasn't necessary after increasing default polling rate to ever 100ms.  Some experimental agents currently use a lower polling rate, but can use the controls provided through an upcoming IOGroup #1914 to set the rate as needed.